tools/autograd/gen_autograd_functions.py

```diff
@@ -417,9 +422,14 @@ def gen_autograd_functions_python(
         },
     )
 
-    infos = list(
-        filter(lambda info: info.args_with_derivatives, differentiability_infos)
-    )
+    # get a 1D list of diffinfos, we do not need them to be per FunctionSchema/DispatchKey here
+    # infos with the diff dispatchkeys but the same name will still be in the same shard.
+    diff_info_list = [
+        info
+        for diffinfo_dict in differentiability_infos.values()
+        for info in diffinfo_dict.values()
+    ]
+    infos = list(filter(lambda info: info.args_with_derivatives, diff_info_list))
```

Comment:
any way this could be factored out into a helper and called from here and 378 as well? (assuming it doesn't disappear per Alban's comment)